### ISSUE TITLE
[QUESTION] Is it possible to use Pico-PIO-USB but mux the pins to the USB Phy microUSB connector?

### ISSUE THREAD

**Issue by shreeve** [sekigon-gonnoc/Pico-PIO-USB#106]
The Pico's have a built in USB Phy and a microUSB connector, along with internal logic for handling the USB communications. However, Pico-PIO-USB is so cool, that it would be nice to be able to use it's logic (and the ability to control what it's doing more easily) instead of the USB Phy built in to the Pico.

Based on the Errata 5 workaround for a hardware issue with the Pico's, I can see that they are muxing the DP/DM pins to solve and issue with the hardware. Thus, it looks possible to re-map the DM/DP pins on to the physical connector and bypass the built-in, hardware based, native USB Phy and have it effectively "replaced" with a pure Pico-PIO-USB version.

Is this possible??? If not, I'm stuck having to solder or otherwise connect to different GPIO pins on each board that we prepare and send out to the field. If we could "commandeer" and take over the microUSB connector on the Pico, but drive it with Pico-PIO-USB, then I wouldn't have to do any soldering or otherwise connecting up wires, etc. and it would be awesome. :-)

Thoughts?

**Comment by shreeve** [comment]
@hathach or @sekigon-gonnoc - Do you know if this is possible?

**Comment by hathach** [comment]
@shreeve No I don't think it is possible to mux different DM/DP pin. The usb_muxing is refering to using different PHY but still using the same pins, much like switching the web back-end from azure to aws (phy), while the front-end (dm/dp pin) is still the same.

If muxing is required, you can use addtional muxing IC e.g https://www.adafruit.com/product/5871

![image](https://github.com/sekigon-gonnoc/Pico-PIO-USB/assets/249515/06d89029-0351-428f-8d7f-5fce95f89f7f)

**Comment by fl-kmarston** [comment]
@shreeve I think you might be doing what I want to do - I want to connect to some devices that have USB PL230 or USB CP210x UARTS.  I want a box with a pico and a USB A that I can connect to via Bluetooth or WiFi and be able to use the serial console of the devices - this then means a phone can access configuration menus running on the single board computers that have the USB UARTS

**Comment by shreeve** [comment]
@sekigon-gonnoc / @hathach / @kilograham - Is it now possible with the rp2350 to mux GPIO pins (here controlled by Pico-PIO-USB) to the USB Phy connector?

**Comment by hathach** [comment]
@shreeve the simple anwser is No. The muxing make it posssible to `use DM/DP as general GPIO`, but the other way is not possible i.e `assign other GPIO to USB PHY`. That is because USB DM/DP is differential signals, they require dedicated analog/filter/logic/gate attached to the pins to get it right (like pio-usb). Most vendor will jusst locked that to a few selected pins where you can switch, for pico dm/dp is the only pins

To sum up you can use DP/DM as GPIO and can possible use pio-usb with it. But you cannot connect other GPIO to the USBPHY.

![image](https://github.com/user-attachments/assets/dd069fec-dc22-4a8d-b6f9-ee33c50690b9)

**Comment by shreeve** [comment]
@hathach - Thank you Ha!

**Comment by shreeve** [comment]
@hathach - Just to clarify, then ... when it says "Use the usb DP and DM pins as GPIO pins instead of connecting them to the USB controller", doesn't this mean that the DP and DM pins *inside the USB Phy* can act as GPIO pins? If they can act as GPIO pins, can we not read from and write to them like GPIO pins?

**Comment by hathach** [comment]
I think when using DM/DP as GPIO, you can turn it on/off and read its value like othe normal GPIO. But it is disconnected from USBPHY making USBHY not acccessible.

```                 
              USBPPHY
        mux  /
DM/DP <-----
            \
              GPIO
```

**Comment by fanoush** [comment]
Hello, I think there is a misunderstanding. As I understand it the initial idea is to control builtin Pico(2) microusb port via Pico-Pio-USB (for reason I don't fully understand/appreciate but let's call it 'simplicity'). This is possible by using DP/DM pins as GPIO. This makes the hardware USB PHY unused  which is what I think @shreeve wants.

This also confirmed by quote
> To sum up you can use DP/DM as GPIO and can possible use pio-usb with it. But you cannot connect other GPIO to the USBPHY.

The other thing discussed which is said to be not possible (but not wanted and not requested?) is to configure the builtin USB PHY to use other pins of the Pico.

The only  question I see now is whether the PIO could access those DP/DM GPIO pins or whether PIO is limited only to 'normal' GPIO. For RP2040 I think this is not possible as DP/DM is not listed in any GPIO bank, however for 2350 the DP/DM pins are in the GPIO Bank 1 together with QSPI pins.

[Page 586](https://datasheets.raspberrypi.com/rp2350/rp2350-datasheet.pdf#page=586) **9.2. Changes from RP2040** says
> USB DP/DM pins can be used as GPIO

[Page 591](https://datasheets.raspberrypi.com/rp2350/rp2350-datasheet.pdf#page=591) **Table 646. GPIO Bank 1 Functions**
DP/DM pins are there and can be reconfigured to be used as GPIO, UART and I2C

So if PIO can access bank 1 it should be possible with 2350. However since the USB PHY would be disabled and one of its features is setting up pull up resistors, one possibly needs to add external 1.5k pull up to D+ or D- pins for  acting as  full/low speed device

**Comment by mungewell** [comment]
For clarity; can someone confirm that the muxing of the USB Phy vs GPIO (PIO) function would allow the RP2040/2350 device to switch between USB Host and USB Device roles without rewiring?

This would be **very helpful** for dev boards with a USB-Type C **PLUG**.

There was some question about PD on ticket #87, but even **without** active PD you can use resistor pull-up/down on CCs to force the power role between Sink/Source. _This is used in Type-C to USB Type-A receptacle._

Maybe a self powered Pico could use other GPIO to force CC pull-ups into particular configuration.

Note: in full (active) PD, power roles and data roles are separate, and are negotiated with the connected device.
